Comparison operators

theodosis1989 opened the pull request:
#### What type of PR is this?

<!--
Choose one of the following labels :
- Improvement
- Bug Fix
-->

#### What does this PR do / why is it needed ?

<!--
Describe change being introduced by this PR.
-->

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
-->
Fixes #

#### Other notes for reviewers:

#### Does this PR introduce a user-facing change?
<!--
-->

theodosis1989 committed:
support for filtering on several levels of nested properties

theodosis1989 committed:
fix for supporting and & or in the same filter

gs-hg committed:
initial commit. tests should run now, provided the running mongodb port is specified correctly in pureSetup.txt ( pure native function to start container + acquired port needs to be implemented)

further clean up to dependencies pending

theodosis1989 committed:
WIP: support for GreaterThan and numbers

theodosis1989 committed:
checked in examples of FEs

theodosis1989 committed:
added comment

theodosis1989 committed:
support for greaterthan for mandatory and optional property

gs-hg committed:
add support for < > <= >= operators ( String and Number )

gs-hg committed:
refactored comparison map into PureFunction-> Lambda instead of Class type

dynamicNew is not recommended as it's not type safe, and had some issues in compiled mode

theodosis1989 committed:
Merge branch 'deeply-nested-filter' into comparison-operators

# Conflicts:
#	legend-engine-xt-nonrelationalStore-mongodb-javaPlatformBinding-pure/src/main/resources/core_nonrelational_mongodb_java_platform_binding/test/pureSetup.txt
#	legend-engine-xt-nonrelationalStore-mongodb-pure/src/main/resources/core_nonrelational_mongodb/pureToDatabaseCommand/pureToDatabaseCommand.pure

gs-hg committed:
added native functions to start,stop mongodb container from pure and also execute mongodb commands from json ( to insert test data for now)

seggregated tests with comparison folders ( greaterThan, lessThan, greaterThanEquals), equality/inequality, graphFetch and some nested queries

gs-hg committed:
should be set to allow for java test executions by default, remove for IDE to compile for now

gs-hg committed:
had country as [0..*] which results in test data changes.

TODO: test domains with 0..* as this breaks our currently pureToDatabaseCommand.pure

gs-hg committed:
Merge branch 'mongo-execute-p5-with-testPlan' into comparison-operators

merging in test suite now with several supported functionalities

# Conflicts:
#	legend-engine-xt-nonrelationalStore-mongodb-executionPlan/src/test/resources/mongoData/person.json
#	legend-engine-xt-nonrelationalStore-mongodb-pure/src/main/resources/core_nonrelational_mongodb/pureToDatabaseCommand/pureToDatabaseCommand.pure

theodosis1989 committed:
handle optional and required multiplicities when constructing nested filters

theodosis1989 committed:
Merge remote-tracking branch 'coworker/comparison-operators' into comparison-operators

theodosis1989 committed:
fixed processNot, all tests are running

theodosis1989 committed:
removed the checked in FEs

theodosis1989 committed:
Merge remote-tracking branch 'coworker/mongo-execute-p5' into comparison-operators

# Conflicts:
#	pom.xml

theodosis1989 committed:
updated the snapshot version on the mongodb-executionPlan-test pom.xml

theodosis1989 committed:
updated the README on how to run a query from the PURE IDE

ghub-real commented on legend-engine-pure-ide-light/src/main/java/org/finos/legend/engine/ide/PureIDELight.java:
```diff
@@ -85,6 +85,7 @@ protected MutableList<RepositoryCodeStorage> buildRepositories(SourceLocationCon
                 .with(this.buildCore("legend-engine-xt-elasticsearch/legend-engine-xt-elasticsearch-V7-pure-metamodel", "elasticsearch_seven_metamodel"))
                 .with(this.buildCore("legend-engine-xt-nonrelationalStore-mongodb-pure","nonrelational-mongodb"))
                 .with(this.buildCore("legend-engine-xt-nonrelationalStore-mongodb-javaPlatformBinding-pure","nonrelational-mongodb-java-platform-binding"))
+                .with(this.buildCore("legend-engine-xt-nonrelationalStore-mongodb-executionPlan-test", "mongodb_execution_test"))
```
i think we should check with elastic - this is not something we want to push out... but only for local development

theodosis1989 replied:
Maybe this is needed only when running the tests from the IDE and we don't need to check it in

gs-hg replied:
Removed this now, added to readme steps to adding this back if and when developers want to run the tests via Legend/Pure IDE

ghub-real commented on legend-engine-xt-nonrelationalStore-mongodb-executionPlan-test/src/main/java/org/finos/legend/engine/plan/execution/stores/mongodb/test/shared/MongoDBCommands.java:
```diff
@@ -0,0 +1,151 @@
+
+        System.out.println(document.toJson());
+
+        return "";
```
sure, we want to return empty string?

theodosis1989 replied:
I'm a little confused with this request method, this method has only one usage and the output seems to not be used. The usage of it is in the java compiled version of _function <<access.private>> meta::external::store::mongodb::executionTest::v7::collectTest(tag: String[1]): PureTestCollection[1]_. There the variable _requestOutput_ is not used

gs-hg replied:
The return is not used in the IDE right now, we have followed what Elastic has done for their request command. For now am returning the JSON (ie. OK )

ghub-real replied:
but we still seem to execute the query - that looks odd - as we will be running it from our executor??

gs-hg replied:
Elastic is also executing queries ( inserting test data) via pure, if you check the pure function call

meta::external::store::elasticsearch::executionTest::testCase::startESCluster

ghub-real commented on legend-engine-xt-nonrelationalStore-mongodb-executionPlan-test/src/main/resources/core_mongodb_execution_test/mongodb_execution_tests.pure:
```diff
@@ -0,0 +1,488 @@
+
+function <<paramTest.BeforePackage>> meta::external::store::mongodb::executionTest::testCase::startMongoDB(config:TestConfig[1]):Boolean[1]
+{
+    println(' meta::external::store::mongodb::executionTest::testCase::startMongoDB');
```
we don't use this? right

gs-hg replied:
good catch, we do this instead in the collect tests portion. removed.

ghub-real commented on legend-engine-xt-nonrelationalStore-mongodb-executionPlan-test/src/test/java/org/finos/legend/engine/plan/execution/stores/mongodb/test/Test_Pure_MongoDB_ExecutionPlan.java:
```diff
@@ -0,0 +1,50 @@
+        TestSuite suite = new TestSuite();
+        if (DockerClientFactory.instance().isDockerAvailable())
+        {
+            System.setProperty("mongo_pwd", "sa");
```
can we review this??  don't want to put this out there, can be a config that needs to be set.  We can add appropriate logging to indicate that some palceholder value is being used, (like PASSWORD_TO_CHANGE)

gs-hg replied:
Looking at PureTestHelper they seem to do this in many other places for tests, we can certainly add this to a config. Elastic equivalent uses the same pattern as well 

we can try to do something else if we prefer but would need to inject at some point

ghub-real replied:
the test container probably has the same value to configure authentictaion??
What about using the same environment variable that engine-authentication uses? (that we set on startup?)

ghub-real commented on pom.xml:
```diff
@@ -310,6 +310,7 @@
         <maven.compiler.target>1.8</maven.compiler.target>
         <maven.compiler.release>8</maven.compiler.release>
         <maven.enforcer.requireJavaVersion>[11.0.10,12)</maven.enforcer.requireJavaVersion>
+        <project.build.outputTimestamp>1680115922</project.build.outputTimestamp>
```
this looks like some unwanted line ... how did we get this?

theodosis1989 replied:
not sure, on the git blame it says it was added on 29/3 by the finos administrator, but now it's not on finos master anymore. I'll look into it

theodosis1989 replied:
it seems that they used to add the ouputTimestamp, but now they stopped. This is commit they added it https://github.com/finos/legend-engine/commit/d118dc020a2b1d030ff4910d1d67e3ad6aa8efc1. When we merge with the latest master it will go away

gs-hg committed:
moved expected test data results to separate resources folder

moved all test setup files to separate resources folder

removed associated maven deps and pure light configs to expose mongodb test data module ( be default we can only run the mongodb execution tests via java build, hidden in the Pure IDE )

cleanup mongodb commands logging

theodosis1989 committed:
WIP: first pass at associations, can get result back for an association with the top level domain class

gs-hg committed:
moved platform binding function ( for tests ) to mongodb-javaPlatformBinding module

Added this function to coreExtensions.pure overrides so we can run tests via Pure IDE or through Java

Cleaned up dependencies not being used and added README.md to execution plan tests and removed binding issue in our executionPlan README.md

gs-hg committed:
cleanup

ghub-real commented on legend-engine-pure-code-compiled-core-configuration/src/main/resources/core_configuration/coreExtensions.pure:
```diff
@@ -83,7 +83,8 @@ function <<LocalPlatformBinding.TestPlanBinder>> meta::pure::extension::configur
       overrides = [
          meta::external::format::flatdata::executionPlan::platformBinding::legendJava::bindFlatdataTestPlanToPlatform__LocalPlatformBinder_1_,
-         meta::relational::executionPlan::platformBinding::legendJava::bindRelationalTestPlanToPlatform__LocalPlatformBinder_1_
+         meta::relational::executionPlan::platformBinding::legendJava::bindRelationalTestPlanToPlatform__LocalPlatformBinder_1_,
```
this was the exact change that we reverted in one of our previous PR.  adding mongo to pure-code-compiled-core-configuration.  Guess this is the binder fix, but let us talk Rafael & Hardik on this.

gs-hg replied:
As discussed, correct, will send out an email to reach consensus. Also the reason I was keen to run a pipeline build prior to merging this into Finos as it has given us issues in the past.

ghub-real commented on legend-engine-xt-nonrelationalStore-mongodb-executionPlan-test/src/main/java/org/finos/legend/engine/plan/execution/stores/mongodb/test/interpreted/MongoDBTestInterpretedExtension.java:
```diff
@@ -0,0 +1,39 @@
+        super(Lists.mutable.with(
+                Tuples.pair(MongoDBCommands.STOP_SERVER_FUNCTION,    MongoDBStopCommand::new),
+                Tuples.pair(MongoDBCommands.START_SERVER_FUNCTION,    MongoDBStartCommand::new),
+                Tuples.pair(MongoDBCommands.REQUEST_SERVER_FUNCTION,    MongodBRequestCommand::new)
```
we are using only start & stop, guess this requestcommand is the one that we copied over, that returns an empty string??

gs-hg replied:
We are using all 3 currently

ghub-real commented on legend-engine-xt-nonrelationalStore-mongodb-executionPlan-test/src/main/resources/core_mongodb_execution_test/mongodb_execution_tests.pure:
```diff
@@ -0,0 +1,449 @@
+
+  let testDataToInsert = meta::pure::functions::io::readFile('/core_mongodb_execution_test/test_setup/person.json')->toOne();
+
+  meta::external::store::mongodb::executionTest::utils::requestMongoDBTestServer($tag, $testDataToInsert);
```
is this the new executeInDB() - is this how we want to insert data?  Thought we moved away from that,  If ES is doing it, let us keep it that way, but I thought the test data inserts is part of TestConnection (runtime) which is where you pass in data to insert?  (that is how it works in studio)

gs-hg replied:
This does seem like a similar pattern that they had identified as anti-pattern for their test setup but it is used in the latest elastic search store execution test modules, as previously mentioned will send out a mail for us to reach consensus

ghub-real commented on legend-engine-xt-nonrelationalStore-mongodb-executionPlan-test/src/main/resources/core_mongodb_execution_test/test_setup/testSetupDSL.txt:
```diff
@@ -0,0 +1,98 @@
+###MongoDB
+Database test::mongodb::mydatabase
```
Suppose this can't be a .pure file, as it would try to compile & fail?  (as it doesn't use legend compile?)

theodosis1989 replied:
A pure file won't allow us to add the new "section headers" like ###MongoDB

ghub-real commented on legend-engine-xt-nonrelationalStore-mongodb-executionPlan-test/src/main/java/org/finos/legend/engine/plan/execution/stores/mongodb/test/shared/MongoDBCommands.java:
```diff
@@ -0,0 +1,136 @@
+        return jsonResult;
+    }
+
+    private static Integer getPortForRunningContainerImage(String imageTag)
```
this seems to be the main value proposition for us, if I understand right.

gs-hg replied:
this and inserting data, yes

theodosis1989 committed:
WIP: nested associations working, need to test more scenarios and cleanup

theodosis1989 committed:
property mappings for associations

theodosis1989 committed:
updated tests with a mix of embedded properties and associations

theodosis1989 committed:
changed associationMappings fetching to a lambda

theodosis1989 committed:
Merge remote-tracking branch 'coworker/comparison-operators' into associations

theodosis1989 committed:
add ignore for warnings in mongodb-javaPlatformBinding-pure

ghub-real committed:
Merge pull request #22 from ghub-real/associations

Associations

ghub-real commented on legend-engine-xt-nonrelationalStore-mongodb-executionPlan-test/src/main/resources/core_mongodb_execution_test/mongodb_execution_tests.pure:
```diff
@@ -0,0 +1,459 @@
+  );
+}
+
+function <<test.TestCollection>> meta::external::store::mongodb::executionTest::v7::test_4_4_1(): PureTestCollection[1]
```
can we check on this one please?  v7 seems like elastic related stuff??
